test/test_images/sendtracker/main.go

```diff
@@ -0,0 +1,150 @@
+		warningLogs = fmt.Sprintf("%sExtra range in received range set\n", warningLogs)
+	}
+	return errorLogs, warningLogs
+}
+
+func main() {
+	sinkURI := os.Getenv("K_SINK")
+	if len(sinkURI) == 0 {
+		fmt.Printf("No SINK URI specified\n")
+		os.Exit(1)
+	}
+	delayMSString := os.Getenv("DELAY_MS")
+	delayMS, err := strconv.Atoi(delayMSString)
+	if err != nil || delayMS < 0 {
+		fmt.Printf("Invalid delay millisecond (%s)", delayMSString)
+		os.Exit(1)
+	}
+	postStopWaitSecsString := os.Getenv("POST_STOP_SECS")
+	postStopWaitSecs, err := strconv.Atoi(postStopWaitSecsString)
+	if err != nil || postStopWaitSecs < 1 {
+		fmt.Printf("Invalid post stop wait seconds (%s)", postStopWaitSecsString)
+		os.Exit(1)
+	}
+	fmt.Printf("Opening client to %s\n", sinkURI)
+	httpOpts := []cehttp.Option{
+		cloudevents.WithTarget(sinkURI),
+	}
+	transport, err := cloudevents.NewHTTP(httpOpts...)
+	if err != nil {
+		fmt.Printf("failed to create transport: %v\n", err)
+		os.Exit(1)
+	}
+
+	ceClient, err := cloudevents.NewClient(transport)
+	if err != nil {
+		fmt.Printf("Unable to create ceClient: %s ", err)
+		os.Exit(1)
+	}
+	fmt.Printf("Client created to %s\n", sinkURI)
+
+	bint, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		fmt.Printf("Error creating stracker random source id: %v", bint)
+		os.Exit(1)
+	}
+	source := fmt.Sprintf("stracker_%s", bint)
+	// Start the receive goroutine
+	receiver := StartReceiver(source, time.Duration(postStopWaitSecs)*time.Second)
+	sendPacer := makeSender(ceClient, receiver, source, time.Duration(delayMS))
+
+	http.HandleFunc("/stop", sendPacer.handleStop)
+	http.HandleFunc("/start", sendPacer.handleStart)
+	http.HandleFunc("/ready", sendPacer.handleReady)
+	http.HandleFunc("/resultsready", sendPacer.handleResultsReady)
+	http.HandleFunc("/results", sendPacer.handleResults)
+	go http.ListenAndServe(":8070", nil)
+
+	// Run the send path synchronously
+	sendError := sendPacer.Run()
+	if sendError != nil && sendPacer.getState() < SendStateFinishing {
+		fmt.Printf("Sender saw an error and never made it to finishing: %v\n", sendError)
+	} else {
+		fmt.Printf("Waiting for receiver to drain\n")
+		<-receiver.fullyDone
+		// Get the events the receiver saw
+		results, dups := receiver.GetResults()
+
+		// Compare with the events the sender sent
+		errorString, warningString := compareLogs(sendPacer.results, results, sendError)
+
+		var resultSummary string
+		// If there were no errors seen, it was a success (warnings don't fail this)
+		if len(errorString) == 0 {
+			resultSummary = "success\n"
+		} else {
+			resultSummary = "failure\n"
+		}
+		sendPacer.resultSummary = fmt.Sprintf("%s%s%sAll send intervals = %s\nAll receive intervals = %s\nDuplicates seen %d\n", resultSummary, errorString, warningString, sendPacer.results, results, dups)
+		fmt.Printf("%s", sendPacer.resultSummary)
+		// Allow the result to be retrieved via /results
+		sendPacer.setState(SendStateDone)
+	}
+
+	// loop forever
```

Comment:
not quite sure why this is needed.

Reply:
I'll add a comment.  Just wanting to keep running forever so that the rest api is still available to get the results out.  Otherwise once we hit the end of main we'd exit the process.

Reply:
Added a comment